crates/accelerate/src/twirling.rs

```diff
@@ -0,0 +1,484 @@
+    ([ZGate, ZGate, IGate, ZGate], 0.),
+    ([XGate, YGate, YGate, ZGate], 0.),
+    ([XGate, IGate, XGate, XGate], 0.),
+    ([XGate, ZGate, YGate, YGate], PI),
+    ([XGate, XGate, XGate, IGate], 0.),
+    ([YGate, IGate, YGate, XGate], 0.),
+    ([YGate, ZGate, XGate, YGate], 0.),
+    ([YGate, XGate, YGate, IGate], 0.),
+    ([YGate, YGate, XGate, ZGate], PI),
+];
+
+static CZ_TWIRL_SET: [([StandardGate; 4], f64); 16] = [
+    ([IGate, ZGate, IGate, ZGate], 0.),
+    ([IGate, XGate, ZGate, XGate], 0.),
+    ([IGate, YGate, ZGate, YGate], 0.),
+    ([IGate, IGate, IGate, IGate], 0.),
+    ([ZGate, XGate, IGate, XGate], 0.),
+    ([ZGate, YGate, IGate, YGate], 0.),
+    ([ZGate, IGate, ZGate, IGate], 0.),
+    ([ZGate, ZGate, ZGate, ZGate], 0.),
+    ([XGate, YGate, YGate, XGate], PI),
+    ([XGate, IGate, XGate, ZGate], 0.),
+    ([XGate, ZGate, XGate, IGate], 0.),
+    ([XGate, XGate, YGate, YGate], 0.),
+    ([YGate, IGate, YGate, ZGate], 0.),
+    ([YGate, ZGate, YGate, IGate], 0.),
+    ([YGate, XGate, XGate, YGate], PI),
+    ([YGate, YGate, XGate, XGate], 0.),
+];
+
+static ISWAP_TWIRL_SET: [([StandardGate; 4], f64); 16] = [
+    ([IGate, ZGate, ZGate, IGate], 0.),
+    ([IGate, XGate, YGate, ZGate], 0.),
+    ([IGate, YGate, XGate, ZGate], PI),
+    ([IGate, IGate, IGate, IGate], 0.),
+    ([ZGate, XGate, YGate, IGate], 0.),
+    ([ZGate, YGate, XGate, IGate], PI),
+    ([ZGate, IGate, IGate, ZGate], 0.),
+    ([ZGate, ZGate, ZGate, ZGate], 0.),
+    ([XGate, YGate, YGate, XGate], 0.),
+    ([XGate, IGate, ZGate, YGate], 0.),
+    ([XGate, ZGate, IGate, YGate], 0.),
+    ([XGate, XGate, XGate, XGate], 0.),
+    ([YGate, IGate, ZGate, XGate], PI),
+    ([YGate, ZGate, IGate, XGate], PI),
+    ([YGate, XGate, XGate, YGate], 0.),
+    ([YGate, YGate, YGate, YGate], 0.),
+];
+
+static TWIRLING_SETS: [&[([StandardGate; 4], f64); 16]; 4] = [
+    &CX_TWIRL_SET,
+    &CZ_TWIRL_SET,
+    &ECR_TWIRL_SET,
+    &ISWAP_TWIRL_SET,
+];
+
+const CX_MASK: u8 = 8;
+const CZ_MASK: u8 = 4;
+const ECR_MASK: u8 = 2;
+const ISWAP_MASK: u8 = 1;
+
+#[inline(always)]
+fn diff_frob_norm_sq(array: ArrayView2<Complex64>, gate_matrix: ArrayView2<Complex64>) -> f64 {
+    let mut res: f64 = 0.;
+    for i in 0..4 {
+        for j in 0..4 {
+            let gate = gate_matrix[[i, j]];
+            let twirled = array[[i, j]];
+            let diff = twirled - gate;
+            res += (diff.conj() * diff).re;
+        }
+    }
+    res
+}
+
+fn generate_twirling_set(gate_matrix: ArrayView2<Complex64>) -> Vec<([StandardGate; 4], f64)> {
+    let mut out_vec = Vec::with_capacity(16);
+    let i_matrix = aview2(&ONE_QUBIT_IDENTITY);
+    let x_matrix = aview2(&qiskit_circuit::gate_matrix::X_GATE);
+    let y_matrix = aview2(&qiskit_circuit::gate_matrix::Y_GATE);
+    let z_matrix = aview2(&qiskit_circuit::gate_matrix::Z_GATE);
+    let iter_set = [IGate, XGate, YGate, ZGate];
+    let kron_set: [Array2<Complex64>; 16] = [
+        kron(&i_matrix, &i_matrix),
```

Comment:
Yeah, that's a good call. I guess the question is should we still do that if the output is otherwise empty? This function will always return 4 identity gates even if there is no other valid twirling combination.

I'm thinking we should probably error if 4 identities are all that matches so only add it to the output list if there are other matches.

Reply:
Yeah an error (or at least a warning) would be good to raise to let users know the code didn't work as expected 👍🏻